packages/atclient/src/atserver_message.c

```diff
@@ -29,7 +33,19 @@ char *atserver_message_get_body(struct atserver_message message) {
 
 static const struct atserver_message EMPTY_MESSAGE = {.buffer = NULL, .len = 0, .prompt_len = 0, .token_len = 0};
 struct atserver_message atserver_message_parse(char *buffer, size_t len) {
-  if (len == 0) {
+  if (len < 1) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Invalid message: not long enough to be a valid server message\n");
+    return EMPTY_MESSAGE;
+  }
+
+  if (len > 2 && buffer[len - 2] == '\r' && buffer[len - 1] == '\n') {
+    buffer[len - 2] = 0;
+    len -= 2;
+  } else if (len > 1 && buffer[len - 1] == '\n') {
+    buffer[len - 1] = 0;
+    len--;
+  } else {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Invalid message: does not end with a newline\n");
```

Comment:
replace the `\r\n` or `\n` with a null terminator, then adjust the string length.

If there is no `\r\n` or `\n` then we reject the message as a failed parse. If parsed directly from the socket layer, assuming the server message is not malformed due to some weird socket / buffer alignment issue (which I've already spent some time on to avoid it happening), the parse should always parse.